### PR TITLE
ARROW-16743: [C++] Add short-circuit version of logical Status AND

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

**Comment by westonpace** [comment]
For an example of the compiler skipping a side-effect: https://godbolt.org/z/bMnnEh4Gn

**Review by cyb70289 — APPROVED** [review]

File: cpp/src/arrow/status.h [hunk 1]
```diff
@@ -424,6 +426,22 @@ Status Status::operator&(Status&& s) const noexcept {
...
 
+Status Status::operator&&(const Status& s) const noexcept {
+  if (ok()) {
+    return s;
+  } else {
+    return *this;
+  }
+}
+
+Status Status::operator&&(Status&& s) const noexcept {
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/src/arrow/status.h:437]: I agree that having a non-reference parameter is preferable. We might still want two overloads, however, since we're copying in the case that `*this` is not ok:

```c++
Status Status::operator&&(Status s) const& noexcept {
  if (!ok()) return *this;
  return s;
}

Status Status::operator&&(Status s) && noexcept {
  if (!ok()) return std::move(*this);
  return s;
}
```

**Comment by pitrou** [comment]
It seems this PR is misled? According to the [C++ docs](https://en.cppreference.com/w/cpp/language/operators) about "rarely overloaded operators":

>The boolean logic operators, `operator&&` and `operator||`. Unlike the built-in versions, the overloads cannot implement short-circuit evaluation.

So -1 from me if my interpretation of this is right. Also cc @bkietz for opinions.

**Comment by bkietz** [comment]
Huh, forgot that wrinkle. Indeed, an overloaded `operator&&` doesn't provide the guarantees of the builtin. Unfortunately, -1

**Comment by bkietz** [comment]
The benefit of hypothetical overridable short-circuiting would be to skip construction of the status on the RHS altogether in the event that the LHS is not ok. That would be neat since we'd be skipping an RHS of arbitrarily complexity with a very terse statement: `Status st = QuickCheck() && TakeFiveMinutesToCheckEntireGraph();`

Those favorable semantics are not possible. The closest we can get is:
```c++
Status st = QuickCheck();
if (st.ok()) {
  st = TakeFiveMinutesToCheckEntireGraph();
}

// or more frequently when we're just returning the status
RETURN_NOT_OK(QuickCheck());
RETURN_NOT_OK(TakeFiveMinutesToCheckEntireGraph());
```

Note that adding member functions like `QuickCheck().And(TakeFiveMinutesToCheckEntireGraph())` would *not* provide short circuiting since we would still always evaluate the expensive RHS.

TL;DR: I don't think there's anything we can fruitfully add here.

**Comment by westonpace** [comment]
Thanks for catching that @pitrou.  I agree there doesn't seem to be any solution here.
